File: internal/tools/read.go
package tools

import (
	"fmt"
	"os"
	"strings"
)

type ReadTool struct{}

func (ReadTool) Name() string {
	return "read"
}

func (ReadTool) Description() string {
	return "Read a file from the filesystem. Returns the file content with line numbers. Supports optional offset and limit for reading specific portions of large files."
}

func (ReadTool) Parameters() map[string]any {
	return objectSchema([]string{"file_path"}, map[string]any{
		"file_path": stringSchema("Absolute path to the file to read"),
		"offset":    numberSchema("Line number to start reading from (1-based)"),
		"limit":     numberSchema("Maximum number of lines to read"),
	})
}

func (ReadTool) Call(input map[string]any) Result {
	filePath, err := stringArg(input, "file_path")
	if err != nil {
		return Result{Output: "Error reading file: " + err.Error(), IsError: true}
	}
	offset, err := optionalNumberArg(input, "offset", 0)
	if err != nil {
		return Result{Output: "Error reading file: " + err.Error(), IsError: true}
	}
	limit, err := optionalNumberArg(input, "limit", 0)
	if err != nil {
		return Result{Output: "Error reading file: " + err.Error(), IsError: true}
	}

	content, err := os.ReadFile(filePath)
	if err != nil {
		return Result{Output: "Error reading file: " + err.Error(), IsError: true}
	}

	lines := strings.Split(string(content), "\n")
	startLine := 0
	if offset > 0 {
		startLine = offset - 1
	}
	if startLine > len(lines) {
		startLine = len(lines)
	}
	endLine := len(lines)
	if limit > 0 && startLine+limit < endLine {
		endLine = startLine + limit
	}

	var builder strings.Builder
	for i, line := range lines[startLine:endLine] {
		if i > 0 {
			builder.WriteByte('\n')
		}
		fmt.Fprintf(&builder, "%6d\t%s", startLine+i+1, line)
	}

	return Result{Output: builder.String()}
}

func (ReadTool) RenderToolCall(input map[string]any) string {
	filePath, _ := stringArg(input, "file_path")
	offset, _ := optionalNumberArg(input, "offset", 0)
	limit, _ := optionalNumberArg(input, "limit", 0)

	description := toolName("read") + " " + fileName(filePath)
	if offset > 0 || limit > 0 {
		parts := []string{}
		if offset > 0 {
			parts = append(parts, fmt.Sprintf("from line %d", offset))
		}
		if limit > 0 {
			parts = append(parts, fmt.Sprintf("limit %d", limit))
		}
		description += muted(" (" + strings.Join(parts, ", ") + ")")
	}
	return description
}

func (ReadTool) RenderResult(result Result) string {
	if result.IsError {
		return failure(result.Output)
	}
	return truncateLines(result.Output, 20, "more lines")
}

File: kai/tools/read.py
from __future__ import annotations

from typing import Any

from kai.tools.base import (
    Result,
    Tool,
    number_schema,
    object_schema,
    optional_number_arg,
    string_schema,
)
from kai.tools.colors import failure, file_name, muted, tool_name
from kai.tools.format import truncate_lines


class ReadTool(Tool):
    def name(self) -> str:
        return "read"

    def description(self) -> str:
        return "Read a file from the filesystem. Returns the file content with line numbers. Supports optional offset and limit for reading specific portions of large files."

    def parameters(self) -> dict[str, Any]:
        return object_schema(
            ["file_path"],
            {
                "file_path": string_schema("Absolute path to the file to read"),
                "offset": number_schema("Line number to start reading from (1-based)"),
                "limit": number_schema("Maximum number of lines to read"),
            },
        )

    def call(self, input: dict[str, Any]) -> Result:
        file_path = input.get("file_path")
        if not isinstance(file_path, str) or not file_path:
            return Result(
                output="Error reading file: missing required argument 'file_path'",
                is_error=True,
            )

        offset = optional_number_arg(input, "offset", 0)
        limit = optional_number_arg(input, "limit", 0)

        try:
            with open(file_path) as f:
                content = f.read()
        except Exception as e:
            return Result(output=f"Error reading file: {e}", is_error=True)

        lines = content.split("\n")
        start = min(offset - 1, len(lines)) if offset > 0 else 0
        end = start + limit if limit > 0 else None
        return Result(
            output="\n".join(
                f"{line_no:>6}\t{line}"
                for line_no, line in enumerate(lines[start:end], start + 1)
            )
        )

    def render_tool_call(self, input: dict[str, Any]) -> str:
        file_path = input.get("file_path", "")
        offset = optional_number_arg(input, "offset", 0)
        limit = optional_number_arg(input, "limit", 0)

        description = f"{tool_name('read')} {file_name(file_path)}"
        if offset > 0 or limit > 0:
            parts = []
            if offset > 0:
                parts.append(f"from line {offset}")
            if limit > 0:
                parts.append(f"limit {limit}")
            description += muted(f" ({', '.join(parts)})")
        return description

    def render_result(self, result: Result) -> str:
        if result.is_error:
            return failure(result.output)
        return truncate_lines(result.output, 20, "more lines")
